activerecord/lib/active_record/connection_adapters/mysql/schema_statements.rb

```diff
@@ -35,13 +35,39 @@ def indexes(table_name)
                 ]
               end
 
-              indexes.last[-2] << row[:Column_name]
-              indexes.last[-1][:lengths][row[:Column_name]] = row[:Sub_part].to_i if row[:Sub_part]
-              indexes.last[-1][:orders].merge!(row[:Column_name] => :desc) if row[:Collation] == "D"
+              if row[:Expression]
+                expression = row[:Expression]
+                expression = +"(#{expression})" unless expression.start_with?("(")
+                indexes.last[-2] << expression
+                indexes.last[-1][:expressions] ||= {}
+                indexes.last[-1][:expressions][expression] = expression
+                indexes.last[-1][:orders][expression] = :desc if row[:Collation] == "D"
+              else
+                indexes.last[-2] << row[:Column_name]
+                indexes.last[-1][:lengths][row[:Column_name]] = row[:Sub_part].to_i if row[:Sub_part]
+                indexes.last[-1][:orders][row[:Column_name]] = :desc if row[:Collation] == "D"
+              end
             end
           end
 
-          indexes.map { |index| IndexDefinition.new(*index) }
+          indexes.map do |index|
+            options = index.last
+
+            if expressions = options.delete(:expressions)
+              orders = options.delete(:orders)
+              lengths = options.delete(:lengths)
+
+              columns = index[-2].map { |name|
+                [ name.to_sym, expressions[name] || +quote_column_name(name) ]
```

Comment:
Why would we want to dup `quote_column_name(name)` here?

Reply:
This is because the quoted columns would be mutated by `add_index_length` and `add_index_sort_order`:

https://github.com/rails/rails/blob/a3dcba42e2422eb9c2e77011a39ce72dc934b420/activerecord/lib/active_record/connection_adapters/mysql/schema_statements.rb#L109-L112

https://github.com/rails/rails/blob/a3dcba42e2422eb9c2e77011a39ce72dc934b420/activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb#L1195-L1198

Reply:
Ah I see, I thought `quote_column_name` would return something that was already duped somehow... because it is the case of https://github.com/rails/rails/blob/61490805fe25091b080f5bba0e286716cdb56245/activerecord/lib/active_record/connection_adapters/abstract/quoting.rb#L67-L69
... but potentially not the case of ...
https://github.com/rails/rails/blob/61490805fe25091b080f5bba0e286716cdb56245/activerecord/lib/active_record/connection_adapters/postgresql/quoting.rb#L46-L48
... got it :)

I'm just realising that there's no way (AFAIK) to know whether a method returns an object that is referenced elsewhere or not, which is kind of sad, because we may be duplicating strings for no reason, maybe we should do something about that one day.

Reply:
Nevermind that's String#-@, and in this case we do know it's referenced elsewhere. I'm blind.